File: service/lib/agama/storage/config_checkers/boot.rb
# frozen_string_literal: true

# Copyright (c) [2024-2025] SUSE LLC
#
# All Rights Reserved.
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of version 2 of the GNU General Public License as published
# by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
# more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, contact SUSE LLC.
#
# To contact SUSE LLC about this file by physical or electronic mail, you may
# find current contact information at www.suse.com.

require "agama/storage/config_checkers/base"
require "yast/i18n"

module Agama
  module Storage
    module ConfigCheckers
      # Class for checking the boot config.
      class Boot < Base
        include Yast::I18n

        # @param storage_config [Storage::Config]
        def initialize(storage_config)
          super()

          textdomain "agama"
          @storage_config = storage_config
        end

        # Boot config issues.
        #
        # @return [Array<Issue>]
        def issues
          [
            missing_alias_issue,
            invalid_alias_issue
          ].compact
        end

      private

        # @return [Storage::Config]
        attr_reader :storage_config

        # @return [Boolean]
        def configure?
          storage_config.boot.configure?
        end

        # @return [String, nil]
        def device_alias
          storage_config.boot.device.device_alias
        end

        # Error if a boot device is required and unknown.
        #
        # This happens when the config solver is not able to infer a boot device, see
        # {ConfigSolvers::Boot}.
        #
        # @return [Issue, nil]
        def missing_alias_issue
          return unless configure? && device_alias.nil?

          error(
            _("The boot device cannot be automatically selected"),
            kind: IssueClasses::Config::NO_ROOT
          )
        end

        # @return [Issue, nil]
        def invalid_alias_issue
          return unless configure? && device_alias && !valid_alias?

          # TRANSLATORS: %s is replaced by a device alias (e.g., "boot").
          error(
            format(_("There is no boot device with alias '%s'"), device_alias),
            kind: IssueClasses::Config::NO_SUCH_ALIAS
          )
        end

        # @return [Boolean]
        def valid_alias?
          return false unless device_alias

          storage_config.supporting_partitions.any? { |d| d.alias?(device_alias) }
        end
      end
    end
  end
end

File: web/src/components/storage/shared/boot.ts
/*
 * Copyright (c) [2026] SUSE LLC
 *
 * All Rights Reserved.
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the Free
 * Software Foundation; either version 2 of the License, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, contact SUSE LLC.
 *
 * To contact SUSE LLC about this file by physical or electronic mail, you may
 * find current contact information at www.suse.com.
 */

import { baseName } from "~/components/storage/utils";
import type { Storage as Proposal } from "~/model/proposal";
import type { Storage as System } from "~/model/system";

/** Where a partition set aside for booting is mounted, where it is mounted at all. */
const BOOT_PATHS = ["/boot", "/boot/efi", "/boot/zipl"];

/**
 * What the backend calls a partition set aside for booting. A partition with no
 * file system describes itself by its partition id in words, "BIOS Boot
 * Partition", "EFI System Partition", and that description is the only place the
 * id reaches the page.
 */
const BOOT_DESCRIPTIONS = /boot|efi|prep|zipl/i;

/** Whether a partition is there to start the machine. */
function isBootPartition(device: Proposal.Device): boolean {
  return (
    device.partition?.efi === true ||
    BOOT_PATHS.includes(device.filesystem?.mountPath || "") ||
    BOOT_DESCRIPTIONS.test(device.description || "")
  );
}

/** One partition the machine will start from, and whether it is there yet. */
export type BootPartition = { name: string; size?: number; isNew: boolean };

/**
 * What booting costs a device, read from the plan rather than the configuration.
 *
 * Nothing in the configuration asks for boot partitions: the installer adds
 * them, so they appear in no view of what was asked for, and a device would look
 * as though it hosted the boot loader for free. Comparing the plan against the
 * machine says which are new and which the device already had.
 *
 * A BIOS boot or PReP partition carries no file system and nothing to mount, so
 * nothing about it says "boot" except that the plan adds it and the
 * configuration never asked for it.
 */
function bootPartitionsOf(
  deviceName: string,
  device: System.Device | null,
  staging: Proposal.Device[],
): BootPartition[] {
  const planned = staging.find((candidate) => candidate.name === deviceName);
  const known = new Set((device?.partitions || []).map((partition) => partition.sid));

  return (planned?.partitions || [])
    .filter(
      (partition) =>
        isBootPartition(partition) || (!known.has(partition.sid) && !partition.filesystem),
    )
    .map((partition) => ({
      name: baseName(partition.name),
      size: partition.block?.size,
      isNew: !known.has(partition.sid),
    }));
}

export { bootPartitionsOf };
